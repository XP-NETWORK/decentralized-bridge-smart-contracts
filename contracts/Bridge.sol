// SPDX-License-Identifier: UNLICENSED
pragma solidity >=0.8.19 <0.9.0;

import "hardhat/console.sol";

import "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
import "@openzeppelin/contracts/token/ERC721/utils/ERC721Holder.sol";
import "@openzeppelin/contracts/token/ERC1155/extensions/IERC1155MetadataURI.sol";
import "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";

contract Bridge is ERC721Holder, ERC1155Holder {
    using ECDSA for bytes32;

    mapping(address => bool) public validators;
    mapping(string => address) public signerAddress;
    uint256 private txFees = 0x0;
    uint256 private validatorsConfirmCount = 0;

    event AddNewValidator(address _validator);
    event Lock(
        uint256 nft_token_id,
        string from_chain,
        address from_chain_user_address,
        address from_chain_nft_address,
        string to_chain,
        string to_chain_nft_address,
        string to_chain_user_address,
        string meta_data,
        string uniuqeIdentifier,
        uint256 txFees
    );

    event UnLock(address to, uint256 tokenId, address contractAddr);

    constructor(address[] memory _validators) {
        for (uint256 i = 0; i < _validators.length; i++) {
            validators[_validators[i]] = true;
        }
    }

    function addNewValidator(address _validator) public {
        emit AddNewValidator(address(_validator));
        validators[_validator] = true;
    }

    function lock(
        uint256 nft_token_id,
        string memory from_chain,
        address from_chain_user_address,
        IERC721Metadata from_chain_nft_address,
        string memory to_chain,
        string memory to_chain_nft_address,
        string memory to_chain_user_address,
        string memory meta_data,
        string memory uniuqeIdentifier,
        bytes32 hash,
        bytes memory sig
    ) external payable requireFees {
        require(
            confirmSignature(hash, sig),
            "validator signature is not valid!"
        );
        if (validatorsConfirmCount >= 7) {
            txFees += msg.value;
            emit Lock(
                nft_token_id,
                from_chain,
                from_chain_user_address,
                address(from_chain_nft_address),
                to_chain,
                to_chain_nft_address,
                to_chain_user_address,
                meta_data,
                uniuqeIdentifier,
                msg.value
            );
            from_chain_nft_address.safeTransferFrom(
                address(msg.sender),
                address(this),
                nft_token_id
            );
            validatorsConfirmCount = 0;
        }
    }

    function unLock(
        address to,
        uint256 tokenId,
        IERC721 contractAddr,
        bytes32 hash,
        bytes memory sig
    ) external payable requireFees {
        require(
            confirmSignature(hash, sig),
            "validator signature is not valid!"
        );
        if (validatorsConfirmCount >= 7) {
            emit UnLock(to, tokenId, address(contractAddr));
            contractAddr.safeTransferFrom(address(this), to, tokenId);
            validatorsConfirmCount = 0;
        }
    }

    modifier requireFees() {
        require(msg.value > 0, "Tx Fees is required!");
        _;
    }

    function claimFee(address payable receiver) external {
        uint256 sendAmt = txFees;
        txFees = 0;
        receiver.transfer(sendAmt);
    }

    function confirmSignature(
        bytes32 hash,
        bytes memory sig
    ) internal returns (bool) {
        validatorsConfirmCount += 1;
        return
            validators[ECDSA.recover(ECDSA.toEthSignedMessageHash(hash), sig)];
    }
}
