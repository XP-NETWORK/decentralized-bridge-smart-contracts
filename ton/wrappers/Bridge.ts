export * from '../build/Bridge/tact_Bridge';