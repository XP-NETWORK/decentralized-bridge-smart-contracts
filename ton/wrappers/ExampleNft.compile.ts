import { CompilerConfig } from '@ton/blueprint';

export const compile: CompilerConfig = {
    lang: 'tact',
    target: 'contracts/example_nft.tact',
};
