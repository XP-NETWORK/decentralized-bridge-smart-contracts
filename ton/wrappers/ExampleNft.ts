export * from '../build/ExampleNft/tact_TestnetNftCollection';
