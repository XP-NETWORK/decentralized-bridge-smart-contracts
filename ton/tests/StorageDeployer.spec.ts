import { Bridge } from "../wrappers/Bridge";
import { ContractSystem, testKey } from "@tact-lang/emulator";
import { Address, beginCell, toNano } from "ton-core";
import { sign } from "ton-crypto";
import { NFTStorageDeployer } from "../build/Bridge/tact_NFTStorageDeployer";
import { NFTCollectionDeployer } from "../build/Bridge/tact_NFTCollectionDeployer";

describe('wallet', () => {
    it('should deploy', async () => {

        // Create wallet
        let key = testKey('wallet-key');
        let publicKey = beginCell().storeBuffer(key.publicKey).endCell().beginParse().loadUintBig(256);
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await Bridge.fromInit(publicKey, "TON"));
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, 'Deploy');
        await system.run();

        // Create executor
        // expect(await contract.getPublicKey()).toBe(publicKey);
        // expect(await contract.getWalletId()).toBe(0n);
        // expect(await contract.getSeqno()).toBe(0n);
        console.log(toNano('10').toString());
        console.log(contract.address)
        console.log(await contract.getCollectionDeployer());
        console.log(await contract.getStorageDeployer());

        await contract.send(treasure, { value: toNano('10') }, 'Lock721');


        // Send transfer and check seqno
        // let transfer: Transfer = {
        //     $$type: 'Transfer',
        //     seqno: 0n,
        //     mode: 1n,
        //     amount: toNano(10),
        //     to: treasure.address,
        //     body: null
        // };
        // let transfer1: Transfer = {
        //     $$type: 'Transfer',
        //     seqno: 0n,
        //     mode: 1n,
        //     amount: toNano(10),
        //     to: treasure.address,
        //     body: beginCell().storeBuffer(Buffer.from("Gando")).endCell()
        // };
        // let signature = sign(beginCell().store(storeTransfer(transfer1)).endCell().hash(), key.secretKey);
        // await contract.send(treasure, { value: toNano(1) }, {
        //     $$type: 'TransferMessage',
        //     transfer: transfer1,
        //     signature: beginCell().storeBuffer(signature).endCell(),
        //     key: publicKey
        // });
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
        // expect(await contract.getSeqno()).toBe(1n);

        // // Send empty message
        // await contract.send(treasure, { value: toNano(1) }, 'notify');
        // await system.run();
        // expect(tracker.collect()).toMatchSnapshot();
        // expect(await contract.getSeqno()).toBe(2n);

        // // Send comment message
        // await contract.send(treasure, { value: toNano(1) }, null);
        // await system.run();
        // expect(tracker.collect()).toMatchSnapshot();
        // expect(await contract.getSeqno()).toBe(3n);
    });
});