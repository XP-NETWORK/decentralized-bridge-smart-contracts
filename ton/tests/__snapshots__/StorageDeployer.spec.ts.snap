// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wallet should deploy 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{4DA28EC08002BE5A3ECB0772CC29DE604D21331181FE1F0FF710171983512690F95022A2883_}
 x{544F4E}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQD7-ob3eGMc9BP5XzGAF5NcVZpu19Wi1oQrYRvQShxQzVP6",
          "type": "internal",
          "value": 50000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 9240n,
      },
    ],
  },
]
`;
